cmd/katapult/vms_test.go

```diff
@@ -0,0 +1,699 @@
+package main
+
+import (
+	"context"
+	"testing"
+
+	"github.com/krystal/go-katapult"
+	"github.com/krystal/go-katapult/core"
+	"github.com/stretchr/testify/assert"
+)
+
+type vmPages [][]*core.VirtualMachine
+
+type vmsClient struct {
+	// Defines a ID that should be not found.
+	idNotFound string
+
+	// Defines an FQDN that should be not found.
+	fqdnNotFound string
+
+	// Defines the power state of the VM's.
+	powerStates map[string]bool
+
+	// Defines the organization ID -> vmPages.
+	organizationIDPages map[string]vmPages
+
+	// Defines the organization subdomain -> vmPages.
+	organizationSubdomainPages map[string]vmPages
+}
+
+// Used to toggle the power state and return the old result.
+func (v *vmsClient) togglePowerState(key string, fqdn bool) bool {
+	if v.powerStates == nil {
+		v.powerStates = map[string]bool{}
+	}
+	prefix := "i"
+	if fqdn {
+		prefix = "s"
+	}
+	state, ok := v.powerStates[prefix+key]
+	if !ok {
+		// A VM starts powered up.
+		state = true
+	}
+	v.powerStates[prefix+key] = !state
+	return state
+}
+
+func (v *vmsClient) List(_ context.Context, org core.OrganizationRef, opts *core.ListOptions) (
+	[]*core.VirtualMachine, *katapult.Response, error,
+) {
+	// Defines the pages.
+	var pages vmPages
+	switch {
+	case org.ID != "":
+		pages = v.organizationIDPages[org.ID]
+	case org.SubDomain != "":
+		pages = v.organizationSubdomainPages[org.SubDomain]
+	default:
+		return nil, nil, katapult.ErrNotFound
+	}
+
+	// Get the VM page.
+	if opts.Page > len(pages) {
+		return nil, nil, katapult.ErrNotFound
+	}
+	page := pages[opts.Page-1]
+
+	// Return the items.
+	totalItems := 0
+	for _, v := range pages {
+		totalItems += len(v)
+	}
+	return page, &katapult.Response{Pagination: &katapult.Pagination{
+		CurrentPage: opts.Page,
+		TotalPages:  len(pages),
+		Total:       totalItems,
+		PerPage:     len(page),
+	}}, nil
+}
+
+func (v *vmsClient) ensureFound(ref core.VirtualMachineRef) error {
+	if (ref.FQDN != "" && v.fqdnNotFound == ref.FQDN) || (ref.ID != "" && v.idNotFound == ref.ID) {
+		return katapult.ErrNotFound
```

Comment:
This should be a `core.ErrVirtualMachineNotFound` to match what the go-katapult package returns when looking up Virtual Machines that don't exist.

```suggestion
		return core.ErrVirtualMachineNotFound
```

The error chain for `core.ErrVirtualMachineNotFound` is: 

- `katapult.Err`
  - `katapult.ErrResponse`
    - `katapult.ErrNotFound`
      - `katapult.ErrResourceNotFound`
        - `core.ErrVirtualMachineNotFound`

It's a bit long, but it ensures that it's easy to target error handling at any specific level you might care about.